qleet/utils/circuit.py

```diff
@@ -0,0 +1,136 @@
+import typing
+
+import numpy as np
+import sympy
+
+import cirq
+import qiskit
+import pyquil
+
+from cirq.contrib.qasm_import import circuit_from_qasm
+from cirq.contrib.quil_import import circuit_from_quil
+
+
+def convert_to_cirq(
+    circuit: typing.Union[qiskit.QuantumCircuit, cirq.Circuit, pyquil.Program]
+) -> cirq.Circuit:
+    """
+    Converts any circuit to cirq
+    :param circuit: input circuit in any framework
+    :return: circuit in cirq
+    """
+    if isinstance(circuit, cirq.Circuit):
+        return circuit
+    elif isinstance(circuit, qiskit.QuantumCircuit):
+        return circuit_from_qasm(circuit.qasm())
+    elif isinstance(circuit, pyquil.Program):
+        return circuit_from_quil(str(circuit))
+    else:
+        raise ValueError(
+            f"Expected a circuit object in cirq, qiskit or pyquil, got {type(circuit)}"
+        )
+
+
+def convert_to_qiskit(
+    circuit: typing.Union[qiskit.QuantumCircuit, cirq.Circuit, pyquil.Program]
+) -> qiskit.QuantumCircuit:
+    """
+    Converts any circuit to qiskit
+    :param circuit: input circuit in any framework
+    :return: circuit in qiskit
+    """
+    if isinstance(circuit, cirq.Circuit):
+        return qiskit.QuantumCircuit.from_qasm_str(circuit.to_qasm())
+    elif isinstance(circuit, qiskit.QuantumCircuit):
+        return circuit
+    elif isinstance(circuit, pyquil.Program):
+        raise convert_to_qiskit(convert_to_cirq(circuit))
+    else:
+        raise ValueError(
+            f"Expected a circuit object in cirq, qiskit or pyquil, got {type(circuit)}"
+        )
+
+
+class CircuitDescriptor:
+    """
+    The interface for users to provide a circuit in any framework and visualize it in qLEET.
+    """
+
+    def __init__(
+        self,
+        circuit: typing.Union[qiskit.QuantumCircuit, cirq.Circuit, pyquil.Program],
+        params: typing.List[sympy.Symbol],
+        cost_function: typing.Callable[[np.ndarray], float],
+    ):
+        self._circuit = circuit
+        self._params = params
+        self._cost = cost_function
+
+    @classmethod
+    def from_qasm(cls, qasm_str: str, params, cost_function):
+        """
+        Generate the descriptor from QASM string
+        :param qasm_str: 3-tuple of QASM strings for each part of the circuit
+        :param params: list of sympy symbols which act as parameters
+        :return: The CircuitDescriptor object
+        """
+        cirq_circuit = circuit_from_qasm(qasm_str)
+        return CircuitDescriptor(
+            circuit=cirq_circuit, params=params, cost_function=cost_function
+        )
+
+    @property
+    def parameters(self) -> typing.List[sympy.Symbol]:
```

Comment:
Do cirq directly take `sympy symbols` or do the need to be converted into some another object first? For example, in qiskit they need to be converted into `Parameter` objects, I guess.

Reply:
We will take a Union later, soon as we implement the qiskit parts.